src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -0,0 +1,183 @@
+    producer_params.emplace_back("bootstrap.servers", kafka->brokers());
+    producer_params.emplace_back("security.protocol", kafka->securityProtocol());
+    if (boost::iequals(kafka->securityProtocol(), "SASL_SSL"))
+    {
+        producer_params.emplace_back("sasl.mechanisms", "PLAIN");
+        producer_params.emplace_back("sasl.username", kafka->username());
+        producer_params.emplace_back("sasl.password", kafka->password());
+    }
+
+    auto * conf = rd_kafka_conf_new();
+
+    char errstr[512] = {'\0'};
+    for (const auto & param : producer_params)
+    {
+        auto ret = rd_kafka_conf_set(conf, param.first.c_str(), param.second.c_str(), errstr, sizeof(errstr));
+        if (ret != RD_KAFKA_CONF_OK)
+        {
+            rd_kafka_conf_destroy(conf);
+            LOG_ERROR(log, "Failed to set kafka config `{}` with value `{}` error={}", param.first, param.second, ret);
+            throw Exception("Failed to create kafka config", ErrorCodes::INVALID_CONFIG_PARAMETER);
+        }
+    }
+
+    rd_kafka_conf_set_dr_msg_cb(conf, [](rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * /* opaque */) {
+        static_cast<WriteBufferFromKafka *>(msg->_private)->onDrMsg(msg);
+    });
+
+    producer = RdKafkaPtr(rd_kafka_new(RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr)), rd_kafka_destroy);
+    if (!producer)
+    {
+        // librdkafka will take the ownership of `conf` if `rd_kafka_new` succeeds,
+        // but if it does not, we need to take care of cleaning it up by ourselves.
+        rd_kafka_conf_destroy(conf);
+        throw Exception("Failed to create kafka handle", klog::mapErrorCode(rd_kafka_last_error()));
+    }
+
+    auto topic = RdKafkaTopicPtr(rd_kafka_topic_new(producer.get(), kafka->topic().c_str(), nullptr), rd_kafka_topic_destroy);
+
+    wb = std::make_unique<WriteBufferFromKafka>(log, std::move(topic));
+
+    std::string data_format = kafka->dataFormat();
+    if (data_format.empty())
+        data_format = "JSONEachRow";
+
+    writer = FormatFactory::instance().getOutputFormat(data_format, *wb, header, context);
+    writer->setAutoFlush();
+
+    pollingThread = std::thread([this]() {
+        while (!isFinished)
+        {
+            auto n = rd_kafka_poll(producer.get(), 500 /* ms */);
+            LOG_TRACE(log, "polled {} events", n);
+        }
+    });
+}
+
+void KafkaSink::consume(Chunk chunk)
+{
+    if (!chunk.hasRows())
+        return;
+
+    auto blk = Block(getHeader());
+    blk.setColumns(chunk.getColumns());
+    writer->write(blk);
+}
+
+void KafkaSink::onFinish()
+{
+    isFinished = true;
+    if (pollingThread.joinable())
+        pollingThread.join();
+
+    // Make sure all outstanding requests are transmitted and handled.
+    if (auto err = rd_kafka_flush(producer.get(), -1 /* blocks */); err)
+        throw Exception(std::string("Failed to flush kafka producer: ") + rd_kafka_err2str(err), klog::mapErrorCode(err));
```

Comment:
It is a bad idea to throw in `dtor` which caused program to abort. Just log everything since fundamentally we can't do anything about the errors at this point.